File: two_pointers/3.longest_substring_no_repeat/3-ver-1.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function (s) {
  let start = 0;
  let max = 0;
  if (s.length <= 1) return s.length;
  while (start < s.length - 1) {
    let temp = s[start];
    let nxt = start + 1;
    while (nxt < s.length) {
      if (!temp.includes(s[nxt])) {
        temp += s[nxt];
        nxt++;
      } else break;
    }
    // console.log(temp)
    max = temp.length > max ? temp.length : max;
    start++;
  }
  return max;
};

// i think this is the sliding window problem also;
// but i used other method and it works;
//
// next time, try other method

File: sliding-window/3-Longest-Substr-without-repeat/3-ver1.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:

        # sliding window solution
        # use set because the lookup time is O(1)
        # charSet = set()
        # l = 0
        # res = 0

        # for r in range(len(s)):
        #     while s[r] in charSet:
        #         charSet.remove(s[l])
        #         l += 1
        #     charSet.add(s[r])
        #     res = max(res, r - l + 1)
        # return res

        # my solution
        if len(s) <= 1:
            return len(s)
        maxLen = 0

        for i in range(len(s)-1):

            temp = s[i]
            start = i+1

            while start < len(s) and s[start] not in temp:
                # print(temp)
                temp += s[start]
                start += 1

            maxLen = max(len(temp), maxLen)

        return maxLen
